src/mongo/db/storage/csv_file.cpp

```diff
@@ -0,0 +1,365 @@
+            ifsMetadata.is_open());
+    std::getline(ifsMetadata, metadata);
+
+    _metadata = getMetadata(parseLine(metadata));
+    ifsMetadata.close();
+
+    _ifs.open(_fileAbsolutePath.c_str(), std::ios::in);
+
+    uassert(ErrorCodes::FileNotOpen,
+            "error = {}"_format(getErrorMessage("open", _fileAbsolutePath)),
+            _ifs.is_open());
+}
+
+// caller must ensure that buffer size is greater than or equal to the size of the bsonObject
+// to be returned. If not, it would throw exception (Not enough size in buffer)
+int CsvFileInput::doRead(char* data, int size) {
+    auto bsonObj = readBsonObj();
+
+    if (!bsonObj) {
+        return 0;
+    }
+
+    int nRead = bsonObj->objsize();
+    uassert(200000402,
+            "Buff Size {} bytes too small to contain {} bytes bsonObj"_format(size, nRead),
+            nRead <= size);
+    memcpy(data, bsonObj->objdata(), nRead);
+    return nRead;
+}
+
+void CsvFileInput::doClose() {
+    _ifs.close();
+}
+
+bool CsvFileInput::isOpen() const {
+    return _ifs.is_open();
+}
+
+bool CsvFileInput::isGood() const {
+    return _ifs.good();
+}
+
+bool CsvFileInput::isFailed() const {
+    return _ifs.fail();
+}
+
+bool CsvFileInput::isEof() const {
+    return _ifs.eof();
+}
+
+// assuming that header is returned by parseLine, which means, each of its element is
+// name of the field with typeInfo
+Metadata CsvFileInput::getMetadata(const std::vector<std::string>& header) {
+    constexpr size_t typeNameAbsent = 0;
+    Metadata ret;
+
+    size_t i = 0;
+    for (const auto& field : header) {
+        // Test scenario where / is missing
+        size_t index = field.find('/');
+        size_t len = index != std::string::npos ? field.length() - index - 1 : typeNameAbsent;
+        // covers both case where field does not / at all or typename absent after /
+        uassert(200000405,
+                "Field \"{}\" does not specify typeName. Index: {}"_format(field, i),
+                len > typeNameAbsent);
+
+        // Test/Guard against scenario where typename is blank
+        auto fieldName = field.substr(0, index);
+        auto typeName = field.substr(index + 1, len);
+        CsvFieldType type;
+
+        if (typeName == "int" || typeName == "int32") {
+            type = CsvFieldType::kInt32;
+        } else if (typeName == "int64" || typeName == "long") {
+            type = CsvFieldType::kInt64;
+        } else if (typeName == "double") {
+            type = CsvFieldType::kDouble;
+        } else if (typeName == "bool") {
+            type = CsvFieldType::kBool;
+        } else if (typeName == "oid") {
+            type = CsvFieldType::kOid;
+        } else if (typeName == "date") {
+            type = CsvFieldType::kDate;
+        } else if (typeName == "string") {
+            type = CsvFieldType::kString;
+        } else {
+            uasserted(200000401, "{} type is not supported. Index: {}"_format(typeName, i));
+        }
```

Comment:
Please include more descriptive information to aid debugging.


```suggestion
            uasserted(200000401, "{} type is not supported at {}th field {}"_format(typeName, i, fieldName));
```

Reply:
Done